pkg/streamingpromql/types/stats.go

```diff
@@ -21,6 +24,75 @@ type QueryStats struct {
 	// For example, if a query is running with a step of 30s with a range vector selector with range 45s,
 	// then samples in the overlapping 15s are counted twice.
 	TotalSamples int64
+
+	// TotalSamplesPerStep represents the total number of samples scanned
+	// per step while evaluating a query. Each step should be identical to the
+	// TotalSamples when a step is run as an instant query, which means
+	// we intentionally do not account for optimizations that happen inside the
+	// range query engine that reduce the actual work that happens.
+	TotalSamplesPerStep []int64
+
+	EnablePerStepStats bool
+	timeRange          QueryTimeRange
+
+	memoryConsumptionTracker *limiter.MemoryConsumptionTracker
+}
+
+func NewQueryStats(timeRange QueryTimeRange, enablePerStepStats bool, memoryConsumptionTracker *limiter.MemoryConsumptionTracker) (*QueryStats, error) {
+	qs := &QueryStats{
+		EnablePerStepStats:       enablePerStepStats,
+		timeRange:                timeRange,
+		memoryConsumptionTracker: memoryConsumptionTracker,
+	}
+	if enablePerStepStats {
+		s, err := Int64SlicePool.Get(qs.timeRange.StepCount, qs.memoryConsumptionTracker)
+		if err != nil {
+			return nil, fmt.Errorf("failed to allocate TotalSamplesPerStep: %w", err)
```

Comment:
There's no need to wrap the error here, and in fact we shouldn't, as this message will be returned to the user, and `TotalSamplesPerStep` won't mean anything to them. The error message returned by `Get` will be user-friendly already.